utilities/transactions/write_unprepared_txn.h

```diff
@@ -17,6 +17,42 @@ namespace rocksdb {
 class WriteUnpreparedTxnDB;
 class WriteUnpreparedTxn;
 
+// WriteUnprepared transactions needs to be able to read their own uncommitted
+// writes, and supporting this requires some careful consideration. Because
+// writes in the current transaction may be flushed to DB already, we cannot
+// rely on the contents of WriteBatchWithIndex to determine whether a key should
+// be visible or not, so we have to remember to check the DB for any uncommitted
+// keys that should be visible to us. First, we will need to change the seek to
+// snapshot logic, to seek to max_visible_seq = max(snap_seq, max_unprep_seq).
+// Any key greater than max_visible_seq should not be visible because they are
+// not unprepared, which means that they are committed after snap_seq.
+//
+// When we seek to max_visible_seq, one of these cases will happen:
+// 1. We hit a unprepared key from the current transaction.
+// 2. We hit a unprepared key from the another transaction.
+// 3. We hit a committed key with snap_seq < seq < max_unprep_seq.
```

Comment:
If seq == max_unprep_seq, then it must be an unprepared key, and cannot be committed.

Reply:
Why? Could not it be a committed value to a key that is not touched by our transaction? Such committed value could have any sequence number including max_unprep_seq.

Reply:
If snap_seq < max_unprep_seq, then this implies the current transaction has unprepared keys written out to db (otherwise, we just have snap_seq == max_unprep_seq).

Since max_unprep_seq is just the max of the set of unprep_seqs, and we know the set is non-empty from the above, then this means max_unprep_seq must be in the set of unprep_seqs, so the key must be unprepared.

Reply:
Note that in my comment I was referring to a key "that is not touched" by our transaction. That key could have a committed value with an arbitrary sequence number.